### PR TITLE
Add experimental JPEG 2000 support with OpenJPEG

### PR DESCRIPTION
**Pull request by stweil**
This pull request is for broader tests and further discussion.

### CODE REVIEW

**Comment by boxerab** [comment]
+1 for the PR

**Comment by ahankinson** [comment]
I'm wondering if OpenJPEG support should be enabled with a flag on the `configure` command rather than as an environment variable.

**Comment by ruven** [comment]
> I'm wondering if OpenJPEG support should be enabled with a flag on the configure command rather than as an environment variable.

Yes, I agree. Apart from for development testing purposes, I don't really see the need to be able to switch between codecs in this way. So, it would make more sense for the choice of JPEG2000 codec to be determined at the configure stage.

File: src/OpenJPEGImage.cc [hunk 1]
```diff
@@ -48,15 +48,13 @@ void OpenJPEGImage::openImage() throw (string){
 		Timer timer; timer.start();
 		logfile << "INFO :: OpenJPEG :: openImage() :: started" << endl << flush;
 	#endif
-	
-	string filename = getFileName( currentX, currentY ); // Get file name
- 	updateTimestamp(filename); // Check if our image has been modified
 
-	if(!(fsrc = fopen(filename.c_str(), "rb"))) throw string("ERROR :: OpenJPEG :: openImage() :: failed to open file for reading");	
+	filename = getFileName( currentX, currentY ); // Get file name
```
Comment (ahankinson) [inline src/OpenJPEGImage.cc]: My C++ is rusty, but shouldn't there be a type definition `string` here?

Comment (ahankinson) [inline src/OpenJPEGImage.cc]: Oh, sorry -- apparently this was superceded...

**Comment by stweil** [comment]
I consider OpenJPEG support to be experimental. If there are no Kakadu files available, the build simply produces a server working with OpenJPEG (no need to set an environment variable). If Kakadu files are available, then this must be the default for compatibility reasons and because it is still much faster, but of course it must also be possible to test OpenJPEG. That's why I added the environment variable.

Would it be better to select Kakadu / OpenJPEG via an URL option? Then it would be possible to compare both on the fly (without running two servers or restarting a single server with different environment settings)?

**Comment by ruven** [comment]
> but of course it must also be possible to test OpenJPEG

It's much cleaner to do this by enabling (or disabling) the choice of decoder through the configure script. Sure, at the moment, Kakadu should be the default if available. If you want to test, you can always build a server with each decoder and run them in parallel, allowing you to compare them.

**Comment by stweil** [comment]
@ruven, which coding style do you prefer for iipsrv? I noticed that the new files for OpenJPEG seem to use a different style (indentation with tabs, function brackets at line endings and more differences) and would like to add a patch which fixes the coding style.

**Comment by stweil** [comment]
I rebased the PR and added three patches which fix compiler warnings.

**Comment by stweil** [comment]
Now a patch which formats the code was added.

**Comment by calvinbutcher** [comment]
+1 for this PR

**Comment by stweil** [comment]
@acdha, was your emoji intentional? What is wrong with the pull request?

**Comment by acdha** [comment]
@stweil Sorry – I've been developing an allergy to "+1" comments but I should have stayed quiet about it

**Comment by stweil** [comment]
@ruven, are you planning own code for OpenJPEG support, or do you want to use the code from this PR? If the later: is there anything which should be changed? I'm asking because I have more modifications for that issue in my queue. I can add them here if this is useful.

**Comment by ahankinson** [comment]
Has the environment variable / compile flag issue been addressed?

**Comment by stweil** [comment]
The current code still supports builds which support both Kakadu and OpenJPEG, with preference for Kakadu at run time and possible selection of OpenJPEG via environment. As without my PR, Kakadu needs to be configured explicitly. OpenJPEG is compiled automatically if it is found. The advantage of the current code is that Kakadu users will get Kakadu as usual and all other users will get OpenJPEG automatically. This also ensures best code coverage for the compilation (no code rotting) without the need for multiple builds.

I have new commits in my local tree which add a configure option for OpenJPEG as well. Thus both Kakadu and OpenJPEG have to be enabled during configuration.

**Comment by boxerab** [comment]
Unfortunately, OpenJPEG doesn't have true region decoding, so it will always be quite slow.
True region decoding only decodes code blocks that lie within a region, plus boundary, and ignores everything else.  One solution is to tile the image with relatively small tiles : 256x256 for example. But, there may be artifacts.

**Comment by stweil** [comment]
For whole images, the current OpenJPEG implementation is slow because of incomplete / old code (missing functions regionDecoding and old signature of OpenJPEGImage::getRegion). Depending on the image size, an improvement by a factor of 100 and more can be reached when fixing that.

**Comment by stweil** [comment]
I rebased the PR and added more commits from my local repository (add missing copyright comments, update exception handling, fix some minor issues like typos and include statements).

**Comment by boxerab** [comment]
@stweil nice work. Let's get this merged into master, and then we can worry about performance improvements.

**Comment by ahankinson** [comment]
> @stweil nice work. Let's get this merged into master, and then we can worry about performance improvements.

Is that the best course of action? IMO, master should be stuff that's ready to go into the next release. If this code still needs work shouldn't it be kept separate until it's ready?

**Comment by boxerab** [comment]
@ahankinson good point - could be merged into an openjpeg branch of iipsrv, and merged into master when ready for release.  Of course, this is Ruven's call.

**Comment by stweil** [comment]
@ahankinson, that's not strictly necessary. IMHO it is sufficient to disable the OpenJPEG part by default and add comments about its experimental status. Then the default builds won't get any of the OpenJPEG code, but developers can enable it easily without the need to switch branches.

**Comment by stweil** [comment]
My [latest working branch](https://github.com/stweil/iipsrv/tree/openjpeg-work) now supports region decoding for full images. This improves the performance of iipsrv with OpenJPEG dramatically. With Kakadu, the total request time is 655142 microseconds for a 4952 x 6536 pixel JP2 image. With OpenJPEG, the total request time is 2062564 microseconds for the same image. Thats a ratio of about 1:3.

**Comment by boxerab** [comment]
@stweil  sounds good.  But, I find the 1:3 ratio surprising, since Kakadu has excellent multi-core support, and does code block level decode. Are you sure this makes sense? How large is the actual image?

**Comment by boxerab** [comment]
If we can achieve a 1:3 ratio, then I think a lot of institutions would consider switching over to a completely open source solution.

**Comment by boxerab** [comment]
What version of OpenJPEG are you using ?

**Comment by ruven** [comment]
Thanks @stweil for all the work you've put in to this. It looks close to being ready to merge this into the master branch, but I'm still not convinced by the need for supporting parallel decoders through an environment variable. It would be far simpler and cleaner to select the decoder to use at the ./configure stage. Otherwise it looks good.

**Comment by stweil** [comment]
@boxerab: I used Debian's libopenjp2-7:amd64 2.1.0-2.1+b1. The test images is here: https://www.weilnetz.de/0335.jp2.

@ruven: The latest version (not the one in the PR) needs `./configure --enable-openjpeg`. The test was made with two different iipsrv.fcgi files – no environment settings involved.

**Comment by stweil** [comment]
> But, I find the 1:3 ratio surprising, since Kakadu has excellent multi-core support, and does code block level decode. Are you sure this makes sense?

When I convert the same image to TIFF on the command line, the ratio is similar:

    time kdu_expand -i 0335.jp2 -o /tmp/0335-kdu.tif
    ...
    real	0m1.156s
    user	0m4.084s
    sys	0m0.088s

    time opj_decompress -i 0335.jp2 -o /tmp/0335-ojp.tif
    ...
    real	0m5.269s
    user	0m5.108s
    sys	0m0.152s

Kakadu uses 4 parallel threads of execution on by notebook computer, while OpenJPEG only uses 1.
This results in a 5:1 ratio for the command line conversion from jp2 to tif. So yes, I think that my previous test result for the conversion from jp2 to jpg with iipsrv makes sense. The numbers here also show that Kakadu's parallel processing is the main factor for the speed difference (just compare the user times).

**Comment by boxerab** [comment]
@stweil thanks. This is interesting: So, user time == single threaded performance ? 
Also, what are the specs for your notebook? On newer hardware, Kakadu will perform much better
due to newer SIMD commands, for example AVX and AVX2. 

Also, in practice, doesn't IIPSrv use very large single tile images and decode only a small region for the user? In this case also, Kakadu will significantly outperform, due to true region decode.

Of course, OpenJPEG can also add multi-threading, AVX support and region decode, and then it will have comparable performance.

**Comment by stweil** [comment]
Notebook spec: Lenovo T430s, Intel Core i7-3520M @ 2.9 GHz (supports AVX)

Real time is the time from start to end (the time which you see). User time is the time used for computing in user mode (sum of all active threads), sys time the time used for computing in system mode (sum of all active threads, typically used for I/O). A single threaded kdu_expand needs 3.06 s which is less than the expected 4.084 s + 0.088 s because multi-threading adds computing overhead.

The timings for opj_decompress used the default image included in Debian. I doubt that it uses special compiler optimisations or AVX support, and it is not based on the latest OpenJPEG, so the single threaded timings of 5.269 s (OpenJPEG) with 3.06 s (Kakadu) are only indicators, not the results of a real competition.

**Comment by boxerab** [comment]
Thanks, I am pleasantly surprised about single thread OpenJPEG performance.  Do you have time to test perf for my multi-threaded J2K codec:  https://github.com/GrokImageCompression/grok ?

Interface is the same as OpenJPEG. Also, fast region decoding is implemented at code block level.

**Comment by stweil** [comment]
Here are some more timings (all from `time bin/opj_decompress -i 0335.jp2 -o 0335.tif`). All tests were executed twice to see the variance in the timing results:

    # latest OpenJPEG, compiled with -O2 -march=native -mtune=native
    real	0m5.095s
    user	0m4.976s
    sys	0m0.116s

    real	0m5.222s
    user	0m4.964s
    sys	0m0.200s

    # latest OpenJPEG, compiled with -O3 -march=native -mtune=native
    real	0m4.940s
    user	0m4.752s
    sys	0m0.168s

    real	0m4.992s
    user	0m4.764s
    sys	0m0.152s

    # latest Grok, compiled with -O3 -march=native -mtune=native
    real	0m4.927s
    user	0m17.984s
    sys	0m0.064s

    real	0m4.927s
    user	0m17.964s
    sys	0m0.100s

The good news is that Grok seems to reduce the system time (did you tune I/O?).
The bad news is that Grok uses much more CPU (user) time, just to reach the same total time as OpenJPEG for my test image (which is not representative).

On a server with many clients where each client starts its own JP2 decompressor, the best performance can be achieved with Kakadu running single threaded, the second best would be the single threaded OpenJPEG.

**Comment by boxerab** [comment]
Thanks for running the numbers. But, something is not right here. On windows 7 64 bit, (core i7 3770 ), decompressing your test image, Grok is about 3 times slower than Kakadu, and OpenJPEG is about 3 times slower than Grok.

This is for multi threaded Grok and Kakadu, and single threaded OpenJPEG.

**Comment by stweil** [comment]
That's indeed strange. I only tested on Debian.

**Comment by boxerab** [comment]
Well, Ubuntu is based on Debian... perhaps the Grok build was not optimized ? That is the only explanation I can think of. Or, your gcc version is old, perhaps. I am using C++11 features in the library.

**Comment by stweil** [comment]
The gcc version is 4.9.2 (Debian Jessie), the build used `-O3 -march=native -mtune=native`. I'll repeat the test on another machine with newer gcc.

**Comment by boxerab** [comment]
OK, thanks for spending the time on this. I think my gcc was 5.x when I tested on XUbuntu.

**Comment by stweil** [comment]
Do you just call `cmake . && make` (as written in file `INSTALL`), or do you use more options (which ones)?

**Comment by boxerab** [comment]
CMAKE_BUILD_TYPE has to be set to Release, and SHARED_BUILD should be enabled.
So, only these two flags. 

I used cmake-gui in my case, since I am running XUbuntu.

**Comment by stweil** [comment]
Setting CMAKE_BUILD_TYPE to Release greatly improves the performance. Similar results can be achieved by setting CMAKE_CXX_FLAGS (which I didn't in my previous tests, I only had set CMAKE_C_FLAGS). Replacing the compiler version 4.9.2 by 6.1.1 only has minor effects on the performance. Test results with same test environment as above:

    # cmake -DCMAKE_C_COMPILER=gcc-6 -DCMAKE_CXX_COMPILER=g++-6 -DCMAKE_C_FLAGS="-Wall -Wextra -O3 -march=native -mtune=native" -DCMAKE_CXX_FLAGS="-Wall -Wextra -O3 -march=native -mtune=native"
    real	0m2.207s
    user	0m7.152s
    sys	0m0.080s
    # cmake -DCMAKE_BUILD_TYPE=Release .
    real	0m2.125s
    user	0m7.244s
    sys	0m0.116s

The overhead for multi-threading is still significant, resulting in a large user time.

**Comment by boxerab** [comment]
Cool. So, Grok is only around twice as slow as Kakadu, for this scenario.  I think that is promising.

**Comment by boxerab** [comment]
Are you able to get iipsrv running with Grok ?

**Comment by stweil** [comment]
Results with iipsrv (`http://localhost/iipsrv/iipsrv.fcgi?FIF=/home/stefan/Downloads/0335.jp2&CVT=jpeg`):

    Grok: Total Request Time: 1218489 microseconds
    OpenJPEG: Total Request Time: 1859932 microseconds

While in my first result, Debian's OpenJPEG was used, the new result is based on latest OpenJPEG built with `cmake -DCMAKE_BUILD_TYPE=Release . && make`.

**Comment by calvinbutcher** [comment]
@stweil, hello, may I ask how you compiled grok and iipsrv? I've been trying for a while (having successfully compiled iipsrv with openjpeg) and keep getting a 500 for a jp2. I'm currently trying to debug with gdb but to no avail.

**Comment by stweil** [comment]
I compiled iipsrv with the openjpeg headers and replaced later the shared library for jp2 (libopenjp2.so.7 / libopenjp2.so.2.1.*) by the one from grok for the test.

Build for grok, openjpeg: `cmake -DCMAKE_BUILD_TYPE=Release . && make`.
Build for iipsrv: `./configure'  'CFLAGS=-g -O2 -Wall -Wextra' 'CXXFLAGS=-g -O2 -Wall -Wextra' 'LDFLAGS=-g' '--enable-openjpeg' && make`.

The latest branch (which was used for all tests) is openjpeg-work.

**Comment by calvinbutcher** [comment]
Great, thanks. I'll try that!

**Comment by stweil** [comment]
@calvinbutcher, an error 500 might be caused by missing, not found or incompatible libraries. I get it when I remove libopenjp2.so.7, for example. The apache2 error log will show the reason like this:

    iipsrv.fcgi: error while loading shared libraries: libopenjp2.so.7: cannot open shared object file: No such file or directory

I suggest to have a look on `/var/log/apache2/error.log` (or wherever your error.log is).

**Comment by boxerab** [comment]
@stweil  thanks for testing with Grok.

**Comment by calvinbutcher** [comment]
Actually, the error I get is:

```
[Fri May 27 15:00:31.254299 2016] [fcgid:warn] [pid 30410:tid 139781290006272] [client 127.0.0.1:60876] mod_fcgid: error reading data, FastCGI server closed connection
[Fri May 27 15:00:31.254513 2016] [core:error] [pid 30410:tid 139781290006272] [client 127.0.0.1:60876] End of script output before headers: iipsrv.fcgi
[Fri May 27 15:00:33.257091 2016] [fcgid:error] [pid 30409:tid 139781373872000] mod_fcgid: process /var/www/localhost/fcgi-bin/iipsrv.fcgi(30554) exit(communication error), get signal 6, possible coredump generated
```

Aaron has kindly guided me in the ways of gdb to try and find the fault. However, I've had issues running iipsrv from the CLI and getting it to hook up with the browser (if that's even what I should be doing!). 

This is probably the first time I've done this much C compiling, so forgive me if I'm being dumb.

When you say "replaced later the shared library for jp2 ... by the one from grok" I assume you mean the  libopenjpeg2 .so generated by the grok compile?

**Comment by boxerab** [comment]
@calvinbutcher  you need to be careful about symlinks - I believe libopenjpeg2.so is just a symlink
to libopenjp2.so.7.

**Comment by calvinbutcher** [comment]
@boxerab understood.

**Comment by stweil** [comment]
Is there another error message a little bit earlier? I also get the ones which you cite, but the missing library is reported before them.

You can also try `ldd PATH/iipsrv.fcgi` to see which libraries will be needed. A missing one will be indicated like this:

	libopenjp2.so.7 => not found

Here you see the reason: `libopenjp2.so.7` exists, but is a symbolic link to `libopenjp2.so.2.1.0` which I renamed for this test.

    lrwxrwxrwx 1 root root     15 Apr  7 19:22 /usr/lib/x86_64-linux-gnu/libopenjp2.so -> libopenjp2.so.7
    -rw-r--r-- 1 root root 225984 Apr  7 19:22 /usr/lib/x86_64-linux-gnu/libopenjp2.so.2.1.0~
    lrwxrwxrwx 1 root root     19 Apr  7 19:22 /usr/lib/x86_64-linux-gnu/libopenjp2.so.7 -> libopenjp2.so.2.1.0

**Comment by boxerab** [comment]
Note: I have bumped Grok version to 2.1.1, so if you get the latest version from github, the `.so` version should match openjpeg.

**Comment by calvinbutcher** [comment]
OK, thank you all. I'll look into this and hopefully get back to you with a working iipsrv-grok docker image.

**Comment by calvinbutcher** [comment]
OK, on the `openjpeg-work` branch of https://github.com/stweil/iipsrv.git I get the following after a make (below):

This normally compiles fine. The only things different are:

- Grok commit 'switch to version 2.1.1'
- Build for grok, openjpeg: `cmake -DCMAKE_BUILD_TYPE=Release . && make`
- Build for iipsrv: `./configure' 'CFLAGS=-g -O2 -Wall -Wextra' 'CXXFLAGS=-g -O2 -Wall -Wextra' 'LDFLAGS=-g' '--enable-openjpeg' && make`
- working on `openjpeg-work` branch instead of `openjpeg` for stweil/iipsrv

I removed the first single quote from the .configure for iipsrv. Assuming the resulting syntax is correct (it configured fine) then I'm really not certain what I've done wrong. 

For iipsrv I normally use `./configure --with-openjpeg=<path to compiled lib>`.

```
IIPImage.h:366:19: warning: unused parameter ‘va’ [-Wunused-parameter]
IIPImage.h:366:19: warning: unused parameter ‘r’ [-Wunused-parameter]
IIPImage.h:366:19: warning: unused parameter ‘layers’ [-Wunused-parameter]
IIPImage.h:366:19: warning: unused parameter ‘x’ [-Wunused-parameter]
IIPImage.h:366:19: warning: unused parameter ‘y’ [-Wunused-parameter]
IIPImage.h:366:19: warning: unused parameter ‘w’ [-Wunused-parameter]
IIPImage.h:366:19: warning: unused parameter ‘h’ [-Wunused-parameter]
OpenJPEGImage.cc: In member function ‘virtual RawTile OpenJPEGImage::getRegion(int, int, unsigned int, int, int, int, unsigned int, unsigned int)’:
OpenJPEGImage.cc:362:7: error: ‘res’ was not declared in this scope
   if (res > numResolutions) {
       ^
OpenJPEGImage.cc:372:40: error: ‘res’ was not declared in this scope
   int vipsres = (numResolutions - 1) - res;
                                        ^
OpenJPEGImage.cc:380:40: error: ‘buf’ was not declared in this scope
   process(w, h, x, y, res, layers, -1, buf);
                                        ^
OpenJPEGImage.cc: At global scope:
OpenJPEGImage.cc:350:9: warning: unused parameter ‘ha’ [-Wunused-parameter]
 RawTile OpenJPEGImage::getRegion(int ha, int va, unsigned int r, int layers,
         ^
OpenJPEGImage.cc:350:9: warning: unused parameter ‘va’ [-Wunused-parameter]
OpenJPEGImage.cc:350:9: warning: unused parameter ‘r’ [-Wunused-parameter]
OpenJPEGImage.cc: In member function ‘virtual RawTile OpenJPEGImage::getRegion(int, int, unsigned int, int, int, int, unsigned int, unsigned int)’:
OpenJPEGImage.cc:386:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
make[1]: *** [OpenJPEGImage.o] Error 1
make[1]: Leaving directory `/home/grok/iip/src'
make: *** [all-recursive] Error 1

```

**Comment by boxerab** [comment]
I got the same error. There is a typo in the signature of the OpenJPEG::getRegion method.

`r` should be changed to `res`

**Comment by stweil** [comment]
My fault: I forgot to push my latest commit. Please update your local copy by `git pull` and try `make` again.

**Comment by boxerab** [comment]
cool. Thanks.

**Comment by calvinbutcher** [comment]
Great, that's compiled now..

**Comment by calvinbutcher** [comment]
@stweil , @boxerab Thanks a lot, it works! I'll tidy up my docker images and they should be stable shortly. Thanks again :)

**Comment by calvinbutcher** [comment]
Now building and testing w/ Loris..

**Comment by stweil** [comment]
@calvinbutcher, may I use your docker images on Monday next week at [ELAG 2016](http://elag2016.org/index.php/program/bootcamps-june-6/exit-leaving-tiff-behind-jpeg-2000-in-libraries/)?

**Comment by calvinbutcher** [comment]
@stweil By all means, though I'd keep an eye on the docker hub builds. The openjpeg builds are stable (https://hub.docker.com/r/bdlss/iipsrv-openjpeg-docker/builds/ and https://hub.docker.com/r/bdlss/loris-openjpeg-docker/builds/)

**However,** the iipsrv-openjpeg image relies on an untagged release of stweil/iipsrv. _Only the loris-openjpeg image is stable with tagged releases of OpenJPEG and Loris_.

If you can give me a tagged release of your iipsrv, I can put this in and test over the next day or so. It would simply make it more stable.

Regarding the rest, I'm still having a little teething trouble with the iipsrv-grok docker image but I hope this to be sorted very soon.

The loris-grok image: I still don't know yet if it's going to work. But again, I hope this will be sorted one way or the other in a day or two.

@boxerab a tagged release of Grok 1.0 as it stands would make these, when finished, more stable.
Please also remember that the iipsrv-grok build is relying on an untagged releases stweil/iipsrv.

**Sorry this reply is a bit wordy. In summary**, tagged releases of Grok and stweil/iipsrv would help make these images more stable for use.

**Comment by stweil** [comment]
Here is your tag: [openjpeg-20160529](https://github.com/stweil/iipsrv/tree/openjpeg-20160529).

**Comment by calvinbutcher** [comment]
Nice, thank you :) I'll put that in both openjpeg and grok builds now..

**Comment by calvinbutcher** [comment]
@stweil just a heads up, someone here had some trouble rolling them out with vagrant. I'm using a virtualbox instance of Ubuntu 14.04 desktop. I've not had time to look into why.

**Comment by calvinbutcher** [comment]
OK, iipsrv-openjpeg and loris-openjpeg builds are stable.

I'm still working on the grok builds. Looks like Pillow doesn't like Grok. I just need to strip out the validation/Pillow requirement for iipsrv-grok and it'll be fine. That'll kill the loris-grok build, though.

**Comment by ruven** [comment]
I've done some preliminary testing with your openjpeg-work branch with iipsrv (changing the default tile size to 256 to make it comparable to the Kakadu implementation) and for an image of size 7185x6217, doing tile requests (JTL command), I get similar timing results to those reported above:

OpenJPEG: 428815
Grok: 313610
Kakadu: 112162

So, Grok is around 3x slower, while standard OpenJPEG is 4x slower than Kakadu. Where performance degrades significantly with respect to Kakadu is with larger images. However, for a larger image of size 15016x11741, I get:

OpenJPEG: Crash!
Grok: 2845229
Kakadu: 124437

So, standard OpenJPEG is unusable, whereas Grok (I guess because of the region decoding that @boxerab added) works, but is around 20x slower than Kakadu, which has a similar decode time to smaller images. Note that these are the decode times for the core process() command in the OpenJPEGImage and KakaduImage classes in iipsrv.

And to put all this into perspective, the same 15016x11741 image in TIFF format has a tile decode time of only 2445, so even Kakadu is significantly slower than TIFF by about a factor of 50!

**Comment by stweil** [comment]
OpenJPEG has problems with large images, see for example https://github.com/uclouvain/openjpeg/pull/730. Can I get your test image somewhere?

**Comment by boxerab** [comment]
@ruven  thanks for testing. For large images, what do you think is a good target speed ratio relative to Kakaudu?  Something greater than 1 :)     @calvinbutcher I will create a tag. @stweil Ruven's test image is only 165 MB, so this is below the 1GB cutoff. I did do a fair bit of memory optimization in Grok, so it does use quite a bit less memory than OpenJPEG.

**Comment by stweil** [comment]
> Ruven's test image is only 165 MB, so this is below the 1GB cutof.

That's why I am interested to know what caused the crash (and to fix it, of course).

**Comment by calvinbutcher** [comment]
@boxerab Thanks. If they're going to be used at ELAN, this will make the docker images more stable. 

FYI Loris fails at setup when installing Pillow, which does not seem to want to install with Grok. I'll look further into this ASAP.

**Comment by calvinbutcher** [comment]
@boxerab Loris build fails when it tries to install a Pillow >= 2.4.0 dependency (it tries to install 3.2).

`libImaging/Jpeg2KDecode.c: In function 'j2k_decode_entry': libImaging/Jpeg2KDecode.c:574:5: error: too few arguments to function 'opj_stream_set_user_data' opj_stream_set_user_data(stream, decoder); ^`

Results in:

`[0mSingle threaded build, not installing mp_compile:1 processes [91merror: Setup script exited with error: command 'x86_64-linux-gnu-gcc' failed with exit status 1[0m[91m [0mRemoving intermediate container 62dddf6d93ca The command '/bin/sh -c ./setup.py install' returned a non-zero code: 1
`

I'm looking into it but I'm not really sure I can help from here :(

**Comment by ruven** [comment]
> Can I get your test image somewhere?

This particular one, no. But there's nothing special about it. It was converted from tiff using Kakadu with parameters including a precinct size of 256x256, a code block size of 64x64 and RPCL ordering.

> That's why I am interested to know what caused the crash (and to fix it, of course).

In fact it was more a memory allocation problem than a seg fault type crash. I guess OpenJPEG tried to put the whole image into memory rather than just a region.

**Comment by calvinbutcher** [comment]
@stweil 
https://github.com/bodleian/loris-openjpeg-docker
https://github.com/bodleian/iipsrv-openjpeg-docker
https://github.com/bodleian/iipsrv-grok-docker

The above are now stable with tagged releases of stweil/iipsrv and grok, so they should be reliable. Am still working on the loris-grok build but may be fixed by the end of play today.

**Comment by boxerab** [comment]
@stweil  this [image  ](http://hirise-pds.lpl.arizona.edu/PDS/RDR/ESP/ORB_028000_028099/ESP_028011_2055/ESP_028011_2055_RED.JP2)  will give you similar characteristics to Ruven's test image. And, if you decode a small region and compare Kakadu to the open source codecs, Kakadu is truly orders of magnitude faster.  So, there is still quite a bit of catching up to do.

**Comment by boxerab** [comment]
By the way, folks, I just discovered a bug in Grok's region decoding : the codec was performing the discrete wavelet transform the the entire image, rather than only for the region. This makes a big difference for large images. So, the latest version on master is now significantly faster when decoding a relatively small region inside a large image.

**Comment by boxerab** [comment]
Hi Ruven,

Do you think this PR is ready for merging, or is there more work to be done? 
Just curious. 

I think once it is merged in, we can get people testing it, and this will help identify areas for
improvement, mostly related to higher performance.

Thanks,
Aaron

**Comment by ruven** [comment]
I'm happy to merge this, but if I understood @stweil correctly, this PR is not in fact the latest version:

> @ruven: The latest version (not the one in the PR) needs ./configure --enable-openjpeg. The test was made with two different iipsrv.fcgi files – no environment settings involved.

@stweil, are you ready to update or do a new pull request?

**Comment by stweil** [comment]
@ruven, my latest code is now in PR #74. It adds the `--enable-openjpeg` option for configure and region decoding for full images. The new code is nearly identical to my latest tagged version [openjpeg-20160529](https://github.com/stweil/iipsrv/tree/openjpeg-20160529), but uses the default tile size of 256 and disabled debug code.

**Comment by boxerab** [comment]
I guess this PR could be closed now ?

**Comment by boxerab** [comment]
@calvinbutcher hope you are enjoying the sun and cider. I just fixed a serious concurrency issue in Grok codec; should I move the tag so that docker image gets the bug-fixed version?

**Comment by stweil** [comment]
Closing this PR. It is superseded by PR #74 (which will hopefully be merged soon).
